python/ray/data/tests/test_dataset.py

```diff
@@ -3260,6 +3260,30 @@ def flaky_mapper(x):
         ).take()
 
 
+def test_dataset_name(ray_start_regular_shared, tmp_path):
+    ds = ray.data.range(10, _dataset_name="MyDataset")
+    assert ds._get_name() == "MyDataset"
+    ds = ds.random_shuffle()
+    assert ds._get_name() == "MyDataset|random_shuffle"
+    ds = ds.map(lambda x: x + 2)
+    assert ds._get_name() == "MyDataset|random_shuffle|map"
```

Comment:
This is going to get quite long, and might run into some limits on the size of the task name. How about we just keep the last stage, and perhaps stage number? Like `MyDataset:map:2`?

Reply:
I feel this makes things more complicated: there is no stage number for a dataset (especially for dataset with multiple ancestors), also repeated datasets will have the same name (e.g. ds.repeat(2).map_each_window()). Another option I’m thinking is using uuid instead of name for now so task name is uuid:shuffle_map_0 and let DatasetStats print out the uuid for each dataset so that I know which dataset a uuid is about. This is not as convenient as a human readable name but it's also enough for debugging at this moment. If it's too inconvenient, we can then introduce name as an alias of uuid if user sets explicitly.

Reply:
Uuid is also pretty long and will get clipped. How about just showing the stage name then? Including the full lineage per task name is going to be unavoidably long.

Reply:
I think just last stage name is not unique enough to be useful. If the entire uuid is too long, how about just the first X digits (should be unique enough) + last stage name.

Reply:
Hmm, why would it not be useful? For an optimized pipeline, there will typically be just a couple stages, like a big fused map and shuffle reduce.

Btw, there are very hard limits here. PR_SET_NAME is limited to 16 chars: https://man7.org/linux/man-pages/man2/prctl.2.html
Given this very limited amount of space, we can't even fit something like: `ray::shuffle_map[0,1,0]`.

Reply:
@ericl Agreed that we should try to have shorter dataset names, but it should be noted that you're referring to the short thread name, which is not what's shown by `ps` and `htop` and is not what we set. We [set](https://github.com/ray-project/ray/blob/30000ff8ae27f9b99041f023c23fd37f62978f1a/python/ray/worker.py#L1817-L1825) the long process title name (`/proc/{pid}/cmdline`) using [`setproctitle`](https://pypi.org/project/setproctitle/), which has a hard-coded limit of the kernel's page size (4KiB).

Reply:
Why don't we do the entire set of chained operations up until some threshold (e.g. 80 characters), and if it's over that threshold, we slice off the (post dataset name) operation prefix until we're under that threshold? Should be some simple enough string manipulation and gives us the most informative task name without letting it get too long.
```
# Under threshold.
ray::MyDataset-map-filter-random_shuffle()
# Over threshold.
ray::MyDataset-...-filter-random_shuffle()
```
```python
OP_SEPARATOR = "-"
TASK_NAME_LENGTH_LIMIT = 75  # 80 - len("ray::")

def _get_task_name(self, op_name: str):
    name = self._get_name() + OP_SEPARATOR + op_name
    if len(name) < TASK_NAME_LENGTH_LIMIT:
        return name
    splits = name.split(OP_SEPARATOR)
    filler = "..."
    running_len = len(splits[0]) + len(filler) + 1
    out = []
    for split in splits[:0:-1]:
        running_len += len(split) + 1
        if running_len > TASK_NAME_LENGTH_LIMIT:
            break
        out.append(split)
    out.append(filler)
    out.append(splits[0])
    return OP_SEPARATOR.join(list(reversed(out))
```
This would then be easy to tweak/extend in the future, e.g. maybe we change the `filler` from an ellipsis to a chain of the first letter of each dropped op, `ray::MyDataset-f-m-m-m-random_shuffle()`.

Reply:
Ok, how about this proposal:
We still have the concept of dataset name: the default is the name of last stage that creates it (for read_api, the data source is the default name) but user has the option to set to their own name if they want. 

The task name is `ray::remote_func_name::dataset_name::dataset_uuid[:8]` so even if it's truncated we still see the remote function that the task is executing. This way the task name won't be super long as well.

cc @clarkzinzow @ericl

Reply:
@jjyao Hmm so this would be:
```
# Explicit dataset name
ray::shuffle_map_2::MyDataset::123e4567
# Last op
ray::map_batches::filter::4567a321
# Default provided by datasource
ray::map_batches::some_parquet_bucket_name::4567a321
```
If the primary goal here is to be able to identify the stage, the underlying task of the stage, and the dataset, then this seems good to me. Although I feel like the `::` separator makes it less readable than `|` or `-`, but that might be up to preference.

For the pipelined ML ingest use case, it would be nice if we could include the epoch number so it's easy to see which epoch each task belongs to...

Reply:
> ray::map_batches::filter::4567a321

This one is confusing (is it doing map or filter)? How about we just show the stage number and task number (actually, this is much more important since it identifies which task/stage may be hanging).